File: hw2/UDP/client/UDPClient.py
#
# SimpleEchoUDPClient.py
#

from socket import AF_INET, SOCK_DGRAM, socket
import time

serverName = 'localhost'
serverPort = 12000

clientSocket = socket(AF_INET, SOCK_DGRAM)
clientSocket.bind(('', 0))

print("Client is running on port", clientSocket.getsockname()[1])

while True:
    print('\n<Menu>')
    print('Please input your client option in integer')
    print('Option 1) Convert text to upper case letters')
    print('Option 2) Ask the server what is my ip and port number')
    print('Option 3) Ask the server how many requests it has served so far')
    print('Option 4) Ask the server how long it has been running since it started')
    try:
        option = int(input('Option :: '))

        if option == 1:
            # send 'ASK_TXTCONV, msg'
            # recv ['converted_msg']
            message_cmd = 'ASK_TXTCONV'
            message_text = input('Input lowercase sentence: ')
            clientSocket.sendto((message_cmd + ',' + message_text).encode(), (serverName, serverPort))
            start_time = time.perf_counter()
            modifiedMessage, serverAddress = clientSocket.recvfrom(2048)
            elapsed_time = time.perf_counter() - start_time
            print('Reply from server:', modifiedMessage.decode())
            print(f'RTT = {elapsed_time*1000:.2f} ms')

        elif option == 2:
            # send 'ASK_IP_PORT'
            # recv [xxx.xxx.xxx.xxx, XXXX] which is my ip and port
            message = 'ASK_IP_PORT'
            clientSocket.sendto(message.encode(), (serverName, serverPort))
            start_time = time.perf_counter()
            modifiedMessage, serverAddress = clientSocket.recvfrom(2048)
            elapsed_time = time.perf_counter()-start_time
            modifiedMessage = modifiedMessage.decode()
            punct_loc = modifiedMessage.find(',')
            print(f'Reply from server: client IP = {modifiedMessage[:punct_loc]}, port = {modifiedMessage[punct_loc+1:]}')
            print(f'RTT = {elapsed_time*1000:.2f} ms')

        elif option == 3:
            # send 'ASK_REQ_NUM'
            # recv [xxx] which is count of requests served
            message = 'ASK_REQ_NUM'
            clientSocket.sendto(message.encode(), (serverName, serverPort))
            start_time = time.perf_counter()
            modifiedMessage, serverAddress = clientSocket.recvfrom(2048)
            elapsed_time = time.perf_counter()-start_time
            print('Reply from server: requests served = ', modifiedMessage.decode())
            print(f'RTT = {elapsed_time*1000:.2f} ms')

        elif option == 4:
            # send 'ASK_RUNTIME'
            # recv [xxx] which is run time of the server since it has been started
            message = 'ASK_RUNTIME'
            clientSocket.sendto(message.encode(), (serverName, serverPort))
            start_time = time.perf_counter()
            modifiedMessage, serverAddress = clientSocket.recvfrom(2048)
            elapsed_time = time.perf_counter()-start_time
            print('Reply from server: run time = ', modifiedMessage.decode())
            print(f'RTT = {elapsed_time*1000:.2f} ms')

        elif option == 5:
            message = 'ASK_CONNEND'
            clientSocket.sendto(message.encode(), (serverName, serverPort))
            print('Bye bye~')
            break

        else:
            print('Wrong option Selected. Please Select menu again.')

    except KeyboardInterrupt:
        message = 'ASK_CONNEND'
        clientSocket.sendto(message.encode(), (serverName, serverPort))
        print('Bye bye~')
        break
    
    except ValueError:
        print('Wrong Input, Please Input Again.')

clientSocket.close()

File: hw2/UDP/client/UDPClient.go
/**
 * UDPClient.go
 **/

package main

import (
	"fmt"
	"net"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"
)

func main() {
	serverName := "localhost"
	serverPort := "12000"

	pconn, _ := net.ListenPacket("udp", ":")

	localAddr := pconn.LocalAddr().(*net.UDPAddr)
	fmt.Printf("Client is running on port %d\n", localAddr.Port)

	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-signals
		var message_cmd string = "ASK_CONNEND"
		server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
		pconn.WriteTo([]byte(message_cmd), server_addr)
		pconn.Close()
		fmt.Println("Bye bye~")
		os.Exit(0)
	}()

	for {
		fmt.Println("\n<Menu>")
		fmt.Println("Please input your client option in integer")
		fmt.Println("Option 1) Convert text to upper case letters")
		fmt.Println("Option 2) Ask the server what is my ip and port number")
		fmt.Println("Option 3) Ask the server how many requests it has served so far")
		fmt.Println("Option 4) Ask the server how long it has been running since it started")
		fmt.Print("Option :: ")

		var user_option int
		fmt.Scanln(&user_option)
		buffer := make([]byte, 1024)

		switch user_option {
		case 1:
			var message_cmd, message_text string = "ASK_TXTCONV", ""
			fmt.Print("Input lowercase sentence: ")
			fmt.Scanln(&message_text)
			server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
			pconn.WriteTo([]byte(message_cmd+","+message_text), server_addr)
			start_time := time.Now()
			count, _, _ := pconn.ReadFrom(buffer)
			elapsed := time.Since(start_time)
			fmt.Printf("Reply from server: %s\n", string(buffer[:count]))
			fmt.Printf("RTT = %d ms\n", elapsed.Milliseconds())
		case 2:
			var message_cmd string = "ASK_IP_PORT"
			server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
			pconn.WriteTo([]byte(message_cmd), server_addr)
			start_time := time.Now()
			count, _, _ := pconn.ReadFrom(buffer)
			elapsed := time.Since(start_time)
			punct_loc := strings.Index(string(buffer), ":")
			fmt.Printf("Reply from server: client IP = %s, port = %s\n", string(buffer)[:punct_loc], string(buffer[punct_loc+1:count]))
			fmt.Printf("RTT = %d ms\n", elapsed.Milliseconds())
		case 3:
			var message_cmd string = "ASK_REQ_NUM"
			server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
			pconn.WriteTo([]byte(message_cmd), server_addr)
			start_time := time.Now()
			count, _, _ := pconn.ReadFrom(buffer)
			elapsed := time.Since(start_time)
			fmt.Printf("Reply from server: requests served = %s\n", string(buffer[:count]))
			fmt.Printf("RTT = %d ms\n", elapsed.Milliseconds())
		case 4:
			var message_cmd string = "ASK_RUNTIME"
			server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
			pconn.WriteTo([]byte(message_cmd), server_addr)
			start_time := time.Now()
			count, _, _ := pconn.ReadFrom(buffer)
			elapsed := time.Since(start_time)
			fmt.Printf("Reply from server: run time = %s\n", string(buffer[:count]))
			fmt.Printf("RTT = %d ms\n", elapsed.Milliseconds())
		case 5:
			var message_cmd string = "ASK_CONNEND"
			server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort)
			pconn.WriteTo([]byte(message_cmd), server_addr)
			fmt.Println("Bye bye~")
			pconn.Close()
			os.Exit(0)
		}
	}
}
